planet/Objects.ocd/Environment.ocd/Cloud.ocd/Script.c

```diff
@@ -249,28 +280,136 @@ private func Precipitation()
 	return;
 }
 
-// Raindrop somewhere from the cloud.
-private func RainDrop()
+// Check if liquid is maybe in frozen form.
+private func RainMat()
 {
-	// Find Random Position.
-	var con = GetCon();
-	var wdt = GetDefWidth() * con / 500;
-	var hgt = GetDefHeight() * con / 700;
-	var x = RandomX(-wdt, wdt);
-	var y = RandomX(-hgt, hgt);
-	if (!GBackSky(x, y))
-		return false;
-	// Check if liquid is maybe in frozen form.
-	var mat;
 	if (rain_mat_freeze_temp != nil && GetTemperature() < rain_mat_freeze_temp)
-		mat = rain_mat_frozen;
+		return rain_mat_frozen;
 	else
-		mat = rain_mat;
-	// Create rain drop.	
-	CastPXS(mat, 1, 1, x, y);
+		return rain_mat;
+}
+
+local last_raindrop_color = -1;
+local particle_cache;
+
+// Raindrop somewhere from the cloud.
+private func RainDrop()
+{
+	var count = Max(rain_visual_strength, 1);
+	for (var i = 0; i < count; i++)
+	{
+		// Find Random Position.
+		var con = GetCon();
+		var wdt = GetDefWidth() * con / 500;
+		var hgt = GetDefHeight() * con / 700;
+		var x = RandomX(-wdt, wdt);
+		var y = RandomX(-hgt, hgt);
+		var xdir = RandomX(GetWind(0,0,1)-5, GetWind(0,0,1)+5)/5;
+		var ydir = 30;
+		if (!GBackSky(x, y))
+			continue;
+		var mat = RainMat();
+
+		if(mat == "Ice")
+		{
+			// Ice (-> hail) falls faster.
+			xdir *= 4;
+			ydir *= 4;
+		}
+
+		// Create rain particles.
+		var particle_name = "Raindrop";
+		var color = GetMaterialColor(mat);
+		if (color != last_raindrop_color)
```

Comment:
The particle definitions stored in the cache only depend on the color and would still work fine if we'd switch between materials of the same color (possibly Lava and DuroLava).